text/0569-tracked-list.md

```diff
@@ -0,0 +1,306 @@
+  }
+}
+```
+
+This pattern is solid and works very well for many Octane users, but it is
+fairly opinionated toward a functional-immutable style of programming, one that
+is different from many Ember apps historically, and that conflicts with
+established programming styles in the ecosystem. This is not necessarily a bad
+thing - one of autotracking's strengths is that it is unopinionated about
+programming style, and can be used with many different paradigms - but it means
+that users who prefer an Object Oriented Programming (OOP) style don't have a
+modern, ergonomic equivalent.
+
+We received feedback to this effect from a good number of Octane early adopters,
+which is why we decided to switch back to using `EmberArray` within the guides
+and documentation for our basic examples.
+
+```js
+import { A } from '@ember/array';
+
+class Person {
+  @friends = A(['Yehuda', 'Melanie', 'Ricardo']);
+
+  addFriend(newFriend) {
+    this.friends.pushObject(newFriend);
+  }
+}
+```
+
+`EmberArray` is reactive, so changes to it will update autotracked values and it
+has the behaviors most OOP users want. However, it has a lot of its own baggage:
+
+- KVO methods like `pushObject` and `popObject` have to be learned in order to
+  use the array, which means learning a whole new array API to be effective in
+  Ember.
+- Custom extra methods like `mapBy` and `sortBy` add to the confusion, since
+  they extend the core array API, and sometimes have conflicting names (e.g.
+  `any` vs `some`)
+- Prototype extensions can still be used, and it can be confusing to new users
+  since their DX is fairly nice, even though they have many downsides and are
+  generally recommended against.
+- Performance is not great, since they either rely on patching the array
+  prototype, which is very bad, or on re-implementing most array methods under
+  the hood.
+
+### Introducing `TrackedList`
+
+The goal of `TrackedList` is to provide a better `EmberArray`, in the short
+term, for Ember users who prefer this style of programming; One that lines up
+more directly with standard array behaviors, and is also reactive.
+
+```js
+import { TrackedList } from '@glimmer/tracking';
+
+class Person {
+  friends = TrackedList.from(['Yehuda', 'Melanie', 'Ricardo']);
+
+  addFriend(newFriend) {
+    this.friends.push(newFriend);
+  }
+
+  get firstFriend() {
+    let [first, ...rest] = this.friends;
+
+    return first;
+  }
+}
+```
+
+`TrackedList` will be as close to 1-1 with the native array API as possible.
+However, it will not attempt to be a drop-in replacement for native arrays,
+since this is not possible without native `Proxy` (which is not supported in
+older browsers) and would have major performance caveats even if it were. This
+is also the motivation behind the name: `TrackedArray` would imply a tracked
+version of native arrays, where `TrackedList` implies that it is its own data
+structure, with potentially divergent behaviors.
```

Comment:
A couple questions as I'm considering this:

- What would those "potentially divergent behaviors be"?
- Is it possible to ship a `TrackedArray` *now*, with future performance improvements unlocked when IE11 is removed (whenever Ember 4 ships), and if so, how do the tradeoffs of doing that compare to the tradeoffs of introducing this class now and then potentially *another* class later?

Reply:
1. Specifically, the two divergences in this RFC are:
  * The difference in behaviors of `sort` and `reverse`
  * The readonly nature of `length`
2. It is not possible to ship `TrackedArray` without native Proxy, so we would have to wait for IE11 support. The native Proxy version is still much much slower than a native array based solution, by at least an order of magnitude, currently.

Reply:
Thanks for the clarifications!